### DISCUSSION TITLE
Setup Webhook

### DISCUSSION THREAD

**Discussion by MasterBehind** [gabrieldwight/Whatsapp-Business-Cloud-Api-Net#60] [General]
Hi Gabriel
Now the solution is working fine from the sense it sends WhatsApp messages, the solution is hosted on https://print-ley.com/whatsapp

Now from Webhook perspective, and from meta configuration to setup a subscription to the Call Back URL, i put the same URL above, but obviously that URL is wrong and it through error as below
“The URL couldn't be validated. Response does not match challenge, expected value="427564319", received="\u003C!DOCTYPE html>\n\u003Ch..."
I believe the URL is incorrect, and it should be the route I have to define in my API project, the question here
How can I define this Route, is it in configuration file? Is it in the code?

**Comment by gabrieldwight** [comment]
in the code on the controller class if you are specifying a custom route using HttpGet/HttpPost attribute or Route attribute

**Reply by MasterBehind** [reply]
I configured it like that [Route("api/webhook")] 
and put the following URL in the Call Back 
https://print-ley.com/whatsapp/api/webhook/
but now it started showing 
The URL couldn't be validated. Callback verification failed with the following errors: HTTP Status Code = 404; HTTP Message = Not Found

knowing that i have published my solution to 
https://print-ley.com/whatsapp

So, what I am missing here?

**Reply by gabrieldwight** [reply]
Test with postman before deploying to see if your route is configured correctly and also check the route configurations for controllers

**Reply by gabrieldwight** [reply]
https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0

**Reply by theCJMan** [reply]
Sample code now (after pull on 2024/03/13) should contain a WebHook Example in the TalkToMeController

**Comment by theCJMan** [comment]
My sincere appologies, I have noticed today that you have a Controller: WhatsAppNotificationController
with [Route("api/[controller]")]
[ApiController]
For the Class and

 [HttpGet("receive/TextMessage")]
  public ActionResult<string> ConfigureWhatsAppMessageWebhook
for the Config of the webHook and

[HttpPost("receive/TextMessage")]
public async Task<IActionResult> ReceiveWhatsAppTextMessage
for receiving messages

Now the very simple controller I added above "TalkToMeController" has 
[HttpGet]
[Route("Home/TalkToMe")]
public ActionResult<string> ConfigureWhatsAppMessageWebhook

I was able to congif my FB WebHook "myappURL/home/talktome" and all worked well, But i want to switch over to this Controller you have in your code..
I am a bit at a loss though, I have tried to replace  to have "myappURL/receive/TextMessage" but it fails...

Please advise

**Reply by gabrieldwight** [reply]
It's the same process you did with talktome route but with different names, which means it has to be verified again from Facebook with the verification code.

**Reply by theCJMan** [reply]
ok, but what is the URL, as it is not working, and I cannot hit it from the out side...

**Reply by gabrieldwight** [reply]
the endpoint you will use to receive webhook results from whatsapp has to be hosted publicly or tested with ngrok or dev tunnel.

**Reply by theCJMan** [reply]
I am using Dev Tunnels but when I am trying to hit your code the resulting page say cannot get to URL, but when I am using my code it is getting through
From the code snippets above, lets say my code is https://www.asdf.com/Home/TalkToMe
Then if I want to switch to your code as below, what should my URL be, I did try https://www.asdf.com/receive/TextMessage

`[Route("api/[controller]")]
[ApiController]
For the Class and

[HttpGet("receive/TextMessage")]
public ActionResult ConfigureWhatsAppMessageWebhook
for the Config of the webHook and

[HttpPost("receive/TextMessage")]
public async Task ReceiveWhatsAppTextMessage
for receiving messages`

**Reply by gabrieldwight** [reply]
Hi, your url should be: https://www.asdf.com/api/[controllerName]/receive/TextMessage

**Comment by theCJMan** [comment]
Maybe I am too new to this Web stuff ;-)
Please note: I have not changed your code so this is under the sample app and 
- the controller file name is: WhatsAppNotificationControl.cs
- the controller class name is: public class WhatsAppNotificationController : ControllerBase
- I have noticed the [Route("api/[controller]")] line in your code
- I have also noticed the line under the above line is [ApiController]

I have now tried the following URLs
https://asdf.com/api/apicontroller/receive/TextMessage                            
    404 Not Found
https://asdf.com/api/WhatsAppNotificationController/receive/TextMessage
    404 Not Found
https://asdf.com/api/apicontroller/WhatsAppNotificationController/receive/TextMessage
    404 Not Found
https://asdf.com/apicontroller/WhatsAppNotificationController/receive/TextMessage 
    404 Not Found
https://asdf.com/WhatsAppNotificationController/receive/TextMessage
    404 Not Found

**Finally** I got https://asdf.com/api/WhatsAppNotification/receive/TextMessage to work

So, I have changed my FB Webhook to point to this address and it was able to authenticate.
Now I am waiting to see if the rest of the `public async Task<IActionResult> ReceiveWhatsAppTextMessage([FromBody] dynamic messageReceived)` work or not..

I have sent a template and a standard Text message from my Temp Number to my personal Cell, and replied from my personal Cell to these messages.
Do you have any idea how long should it take for the "ReceiveWhatsAppTextMessage" to fire?

**Reply by theCJMan** [reply]
BTW, got it working!!!!
Do remember to:
1. On FB after authenticating your webhook
2. THEN you must subscribe to the WebHook fields you want to get notified about...
 
Once you have done BOTH point 1 and 2, then you should receive the messages (as per your field subscription)
